File: 46-permutations/46-permutations.go
func permute(nums []int) [][]int {
    results := make([][]int, 0)
	if len(nums) == 0 {
		results = append(results, []int{})
		return results
	}

	for _, subResult := range permute(nums[1:]) {
		for index := 0; index < len(nums); index++ {
			result := append([]int{}, subResult[:index]...)
			result = append(result, nums[0])
			result = append(result, subResult[index:]...)
			results = append(results, result)
		}
	}
	return results
}

File: 46-permutations/46-permutations.ts

// [1,2,3]
// [1] choice [2,3]
// [1,2] choice [3]
// [1,2,3] choice []

// [2] choice [1,3]
// [2,1] choice [3]
// [2,1,3] choice []

//...
function permute(nums: number[],permutation = [], answer = []) : number[][] {
    if(nums.length === 0){
        answer.push([...permutation])
    }
    for (let i = 0 ; i < nums.length ; i++){
        permutation.push(nums[i]);
        const choices = nums.filter((num,index)=> index !== i);
        permute(choices,permutation,answer)
        permutation.pop()
    }
   
    return answer
};